### PR TITLE
task-4 implementation

### PR DESCRIPTION
**Pull request by AndreyAusianik**
api works as expected
https://3i0mj63la9.execute-api.eu-west-1.amazonaws.com/dev/products/

sql file with creates/inserts is commited as task3.1.sql

FE requests products from lambda (but no changes were made) https://d32wiwybt1hsj3.cloudfront.net/

Additional (optional) tasks:
+1 - POST/products lambda functions returns error 400 status code if product data is invalid
+1 - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
+1 - All lambdas do console.log for each incoming requests and their arguments
+1 - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa)

### CODE REVIEW

File: src/libs/pg-client.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    } catch(err) {
+        console.error('error during database request executing:', err);
+        throw err;
+    } finally {
+        client.end();
+    }
+}
+
+export const executeTransaction  = async (transactions: [string, any[]?][]):Promise<QueryResult[]> => {
+    const client = new Client(dbOptions);
```
Comment (roman-shuvalov) [inline src/libs/pg-client.ts:32]: No need to create a new client and connection for each manipulation with DB. Would be better to create a connections pool in the module scope and use them.

File: src/functions/createProduct/handler.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        ['%RESULT:1:product_id%', product.count]
+      ]
+    ]);
+    console.log('everything worked');
+    return formatJSONResponse({
+      ...product,
+      product_id,
+    });
+  } catch (e) {
+    if(['23502'/*not_null_violation*/, '42804', /*datatype_mismatch*/].includes(e.code)) {
```
Comment (roman-shuvalov) [inline src/functions/createProduct/handler.ts:32]: Before DB would be great to have backend validation.
For example, using ajv package.

File: src/functions/createProduct/handler.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import 'source-map-support/register';
+
+import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/apiGateway';
+import { formatJSONResponse, formatError } from '@libs/apiGateway';
+import { middyfy } from '@libs/lambda';
+import { executeTransaction } from '@libs/pg-client';
+import { Product } from '@models/product';
+
+const getProductsById: ValidatedEventAPIGatewayProxyEvent<null> = async (event) => {
```
Comment (roman-shuvalov) [inline src/functions/createProduct/handler.ts]: Looks like name should be something like **createProduct**

File: src/functions/createProduct/index.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  handler: `${handlerPath(__dirname)}/handler.main`,
+  events: [
+    {
+      http: {
+        method: 'post',
+        path: 'products',
+      }
+    }
+  ]
+}
```
Comment (roman-shuvalov) [inline src/functions/createProduct/index.ts:24]: Post request returns error
Screenshot 2021-08-02 at 18 09 57

File: src/libs/pg-client.ts (hunk 1 above)
Comment (AndreyAusianik) [reply src/libs/pg-client.ts:32]: completely disagree with this point. you don't control when the lambda "starts", and how many node processes you have to controll that. With that in mind this "pool" logic just does not work. Pool thing is rather done for simple servers, which remain turned on from the moment of starting it. During the lection it was explicitly said that for lambdas Client is a better option

File: src/functions/createProduct/handler.ts (hunk 2 above)
Comment (AndreyAusianik) [reply src/functions/createProduct/handler.ts:32]: agree, thanks for a piece of advice. But there was no task to implement that type of validation

File: src/functions/createProduct/index.ts (hunk 4 above)
Comment (AndreyAusianik) [reply src/functions/createProduct/index.ts:24]: thanks. My mistake. it seems I was sending json string during my tests

**Comment by alimovdavron** [comment]
image
5 - Your own Frontend application is integrated with product service (/products API) and products from product-service are represented on Frontend. Link to a working Front-End application is provided for cross-check reviewer. This task is not implemented, I can't add product using FE.

**Comment by AndreyAusianik** [comment]
updated. FE PR: https://github.com/AndreyAusianik/nodejs-aws-fe/pull/3. product form now creates objects

**Pull request merged**